backend/inspirehep/pidstore/providers/texkey.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020 CERN.
+#
+# inspirehep is free software; you can redistribute it and/or modify it under
+# the terms of the MIT License; see LICENSE file for more details.
+import random
+import string
+
+from flask import current_app
+from inspire_utils.record import get_value
+from invenio_pidstore.models import PersistentIdentifier, PIDStatus
+from invenio_pidstore.providers.base import BaseProvider
+
+from inspirehep.pidstore.errors import NoAvailableTexKeyFound, PIDAlreadyExistsError
+from inspirehep.records.utils import get_last_name, get_literature_earliest_date
+
+
+class InspireTexKeyProvider(BaseProvider):
+    pid_type = "texkey"
+    pid_provider = "texkey"
+    default_status = PIDStatus.RESERVED
+
+    @classmethod
+    def sanitize(cls, name):
+        name = name.split()[0].lower()
+        name = "".join(filter(lambda x: x in set(string.ascii_lowercase), name))
+        return name.capitalize()
+
+    @classmethod
+    def get_texkey_first_part(cls, data):
+        if "authors" in data and len(data["authors"]) < 10:
+            return cls.sanitize(get_last_name(data["authors"][0]))
+        elif "collaborations" in data:
+            return cls.sanitize(data["collaborations"][0]["value"])
+        elif "corporate_author" in data:
+            return cls.sanitize(data["corporate_author"][0])
+        elif "proceedings" in data["document_type"]:
+            return cls.sanitize("Proceedings")
+        elif "authors" in data:
+            return cls.sanitize(get_last_name(data["authors"][0]))
+        else:
+            return None
+
+    @classmethod
+    def get_texkey_second_part(cls, data):
+        date = get_literature_earliest_date(data)
+        if date is not None:
+            return date.split("-", 1)[0]
+        return None
+
+    @classmethod
+    def get_texkey_with_random_part(cls, texkey):
+        retry_count = current_app.config.get("PIDSTORE_TEXKEY_MAX_RETRY_COUNT", 5)
+        size = current_app.config.get("PIDSTORE_TEXKEY_RANDOM_PART_SIZE", 3)
+        for _ in range(retry_count):
+            random_part = "".join(
+                random.choices(string.ascii_lowercase + string.digits, k=size)
+            )
+            pid_value = f"{texkey}{random_part}"
+            if not cls.texkey_exists(pid_value):
+                return pid_value
```

Comment:
instead of randomly retry, could we use something like https://invenio-sequencegenerator.readthedocs.io ?

Reply:
I was looking into invenio-sequence generator. It generates only sequences (1,2,3 or A,B,C etc) not random numbers. In specification for TexKey it says specifically that last part should be random.